crates/goose/src/recipe/build_recipe/mod.rs

```diff
@@ -0,0 +1,173 @@
+use crate::recipe::read_recipe_file_content::RecipeFile;
+use crate::recipe::template_recipe::{parse_recipe_content, render_recipe_content_with_params};
+use crate::recipe::{
+    Recipe, RecipeParameter, RecipeParameterRequirement, BUILT_IN_RECIPE_DIR_PARAM,
+};
+use anyhow::Result;
+use std::collections::{HashMap, HashSet};
+
+pub fn render_recipe_template<F>(
+    recipe_file: RecipeFile,
+    params: Vec<(String, String)>,
+    user_prompt_fn: Option<F>,
+) -> Result<(String, Vec<String>)>
+where
+    F: Fn(&str, &str) -> Result<String, anyhow::Error>,
+{
+    let RecipeFile {
+        content: recipe_file_content,
+        parent_dir: recipe_parent_dir,
+        ..
+    } = recipe_file;
+    let recipe_dir_str = recipe_parent_dir
+        .to_str()
+        .ok_or_else(|| anyhow::anyhow!("Error getting recipe directory"))?;
+    let recipe_parameters = validate_recipe_parameters(&recipe_file_content, recipe_dir_str)?;
+
+    let (params_for_template, missing_params) =
+        apply_values_to_parameters(&params, recipe_parameters, recipe_dir_str, user_prompt_fn)?;
+
+    let rendered_content = if missing_params.is_empty() {
+        render_recipe_content_with_params(&recipe_file_content, &params_for_template)?
+    } else {
+        String::new()
+    };
+
+    Ok((rendered_content, missing_params))
+}
+
+pub fn validate_recipe_parameters(
+    recipe_file_content: &str,
+    recipe_dir_str: &str,
+) -> Result<Option<Vec<RecipeParameter>>> {
```

Comment:
`Option<Vec<..>>` could just be Vec which could be empty? any distinction between a None vs an empty Vec here?